nixos/modules/tasks/auto-upgrade.nix

```diff
@@ -91,24 +112,36 @@ let cfg = config.system.autoUpgrade; in
       path = with pkgs; [ coreutils gnutar xz.bin gzip gitMinimal config.nix.package.out ];
 
       script = let
-          nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
-        in
+        nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
+        date = "/run/current-system/sw/bin/date";
+        readlink = "/run/current-system/sw/bin/readlink";
+        shutdown = "/run/current-system/sw/bin/shutdown";
+        upgradeFlag = optional (cfg.channel == null) "--upgrade";
+      in
         if cfg.allowReboot then ''
-            ${nixos-rebuild} boot ${toString cfg.flags}
-            booted="$(readlink /run/booted-system/{initrd,kernel,kernel-modules})"
-            built="$(readlink /nix/var/nix/profiles/system/{initrd,kernel,kernel-modules})"
-            if [ "$booted" = "$built" ]; then
-              ${nixos-rebuild} switch ${toString cfg.flags}
-            else
-              /run/current-system/sw/bin/shutdown -r +1
-            fi
-          '' else ''
+          ${nixos-rebuild} boot ${toString (cfg.flags ++ upgradeFlag)}
+          booted="$(${readlink} /run/booted-system/{initrd,kernel,kernel-modules})"
+          built="$(${readlink} /nix/var/nix/profiles/system/{initrd,kernel,kernel-modules})"
+          ${optionalString (cfg.rebootWindow != null) ''current_time="$(${date} +%H:%M)"''}
+
+          if [ "''${booted}" = "''${built}" ]; then
             ${nixos-rebuild} switch ${toString cfg.flags}
-        '';
+          ${optionalString (cfg.rebootWindow != null) ''
+            elif [[ "''${current_time}" < "${cfg.rebootWindow.lower}" ]] || \
+                 [[ "''${current_time}" > "${cfg.rebootWindow.upper}" ]]; then
+              echo "Outside of configured reboot window, skipping."
```

Comment:
Might be an overkill but a transient timer could just schedule the reboot into the next reboot window: https://wiki.archlinux.org/index.php/Systemd/Timers#Transient_.timer_units

Reply:
Or just `shutdown -r "$cfg.rebootWindow.lower"`.

Reply:
I guess we'd have to introduce a new option for that though. And maybe we'd need to make the days explicit too then, my servers only have a maintenance window  once per week, for instance, not every day...
We might be able to make the whole thing more flexible by integrating with systemd somehow instead of comparing strings ourselves, but I'd have to think a bit about how that could be done.

Reply:
Also, it kind of gets pushed into the next maintenance window already now, given that the service should run again then and still detect the changes between the current system and the new boot default.

Reply:
Ok. Sounds good enough to me than!